### PR TITLE
NP-6365 - Created SuggestedTextField

### PR DESCRIPTION
**Pull request by tala-abuadas**
image

image (1)

### CODE REVIEW

File: src/foam/u2/view/SuggestedTextField.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      .on('focus', () => {
+        this.inputFocused = true;
+      })
+      .on('blur', () => {
+        this.inputFocused = false;
+      })
+      .end()
+      .add(this.slot(function(filteredValues, data, inputFocused) {
+        if ( ! data || ! inputFocused ) return this.E();
+        if ( ! filteredValues.length ) return this.E().addClass(this.myClass("suggestions")).add(this.emptyTitle)
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/SuggestedTextField.js]: single quotes

File: src/foam/u2/Autocompleter.js [hunk 2]
```diff
@@ -44,29 +44,24 @@ foam.CLASS({
...
-      }.bind(this), this.dao$, this.partial$);
+      this.onUpdate();
     }
   ],
 
   listeners: [
     {
       name: 'onUpdate',
-      isFramed: true,
+//      isFramed: true,
```
Comment (sarthak-marwaha) [inline src/foam/u2/Autocompleter.js]: Should this be commented?

File: src/foam/u2/view/SuggestedTextField.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    }
+  ],
+
+  listeners: [
+    {
+      name: 'onUpdate',
+      isFramed: true,
+      code: function() {
+        this.autocompleter.onUpdate();
+        this.autocompleter.filteredDAO.limit(5).select()
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/SuggestedTextField.js]: Shouldn't the limit be on the autocompleter? Also should be configureable

File: src/foam/u2/view/SuggestedTextField.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        this.inputFocused = false;
+      })
+      .end()
+      .add(this.slot(function(filteredValues, data, inputFocused) {
+        if ( ! data || ! inputFocused ) return this.E();
+        if ( ! filteredValues.length ) return this.E().addClass(this.myClass('suggestions')).add(this.emptyTitle)
+        return this.E().addClass(this.myClass('suggestions')).add(this.title).forEach(this.filteredValues, function(obj) {
+          this
+           .start(self.rowView, { data: obj })
+           .addClass(self.myClass('row'))
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/SuggestedTextField.js]: indent

File: src/foam/u2/view/SuggestedTextField.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        this.autocompleter.partial$ = this.data$;
+
+      this.onDetach(this.onload.sub(this.loaded));
+      this
+      .start(this.TextField, {
+        data$: this.data$,
+        onKey: true,
+        placeholder$: this.placeholder$
+      })
+      .on('focus', () => {
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/SuggestedTextField.js]: indent

**Review by sarthak-marwaha — APPROVED** [review]

**Review by jlhughes — COMMENTED** [review]
LGTM

File: src/foam/u2/view/SuggestedTextField.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      }))
+    }
+  ],
+
+  listeners: [
+    {
+      name: 'onUpdate',
+      isFramed: true,
+      code: function() {
+        this.autocompleter.filteredDAO.select()
```
Comment (lchanmann) [inline src/foam/u2/view/SuggestedTextField.js:153]: can filteredDAO be null?

File: src/foam/u2/view/SuggestedTextField.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        .on('focus', () => {
+          this.inputFocused = true;
+        })
+        .on('blur', () => {
+          this.inputFocused = false;
+        })
+      .end()
+      .add(this.slot(function(filteredValues, data, inputFocused) {
+        if ( ! data || ! inputFocused ) return this.E();
+        if ( ! filteredValues.length ) return this.E().addClass(this.myClass('suggestions')).add(this.emptyTitle)
```
Comment (lchanmann) [inline src/foam/u2/view/SuggestedTextField.js]: please add semi-colon at the end.

File: src/foam/u2/view/SuggestedTextField.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/u2/view/SuggestedTextField.js]: 2022

File: src/foam/u2/view/SuggestedTextField.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+ foam.CLASS({
+  package: 'foam.u2.view',
+  name: 'SuggestedTextField',
+  extends: 'foam.u2.View',
+
```
Comment (kgrgreer) [inline src/foam/u2/view/SuggestedTextField.js:10]: Please add documentation:

File: src/foam/u2/view/SuggestedTextField.js [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+      }
+    },
+    {
+      name: 'onSelect',
+      value: function(obj) {
+        this.data = obj.toSummary();
+      }
+    },
+    {
+      name: 'ctx'
```
Comment (kgrgreer) [inline src/foam/u2/view/SuggestedTextField.js]: What is this used for?

File: src/foam/u2/view/SuggestedTextField.js [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+      .end()
+      .add(this.slot(function(filteredValues, data, inputFocused) {
+        if ( ! data || ! inputFocused ) return this.E();
+        if ( ! filteredValues.length ) return this.E().addClass(this.myClass('suggestions')).add(this.emptyTitle)
+        return this.E().addClass(this.myClass('suggestions')).add(this.title).forEach(this.filteredValues, function(obj) {
+          this
+           .start(self.rowView, { data: obj })
+             .addClass(self.myClass('row'))
+             .on('mousedown', function() {
+               self.onSelect.call(self, obj, self.ctx)
```
Comment (lchanmann) [inline src/foam/u2/view/SuggestedTextField.js]: onSelect doesn't use `ctx`, what is the plan passing the context?

File: src/foam/u2/view/SuggestedTextField.js [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+      class: 'foam.u2.ViewSpec',
+      name: 'rowView',
+      value: {
+        class: 'foam.u2.CitationView'
+      }
+    },
+    {
+      name: 'onSelect',
+      value: function(obj) {
+        this.data = obj.toSummary();
```
Comment (lchanmann) [inline src/foam/u2/view/SuggestedTextField.js:78]: can we publish a topic eg. 'selected' that others can subscribe to?

File: src/foam/u2/Autocompleter.js [hunk 13]
```diff
@@ -44,22 +44,19 @@ foam.CLASS({
...
-          'the chosen completion, we need to turn it back into a string for ' +
-          'the text field.',
-      required: true
-    },
-    'filteredDAO'
+    class: 'foam.dao.DAOProperty',
+    name: 'filteredDAO',
+    factory: function() {
+      return this.dao;
+     }
```
Comment (lchanmann) [inline src/foam/u2/Autocompleter.js]: please indent.

**Comment by sarthak-marwaha** [comment]
@lchanmann @kgrgreer RE: ctx, the idea was for the parent element to be able to pass in both the onSelect function and their data$(or any other prop$) in the ctx$ and then the SuggestedTextField would handle calling the function which could use the ctx variable to set values on the parent element from the row object. If there is a better way to do this, then we can use that approach

File: src/foam/u2/view/SuggestedTextField.js (hunk 6 above)
Comment (tala-abuadas) [reply src/foam/u2/view/SuggestedTextField.js:153]: no, the factory sets it to the full dao

**Pull request merged**